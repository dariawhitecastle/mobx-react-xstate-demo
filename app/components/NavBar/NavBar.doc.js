import { describe, PropTypes } from 'react-desc'

export const doc = NavBar => {
  const DocumentedNavBar = describe(NavBar)
    .description('A NavBar')
    .usage(
      `import { NavBar } from 'components/NavBar';
<NavBar />`,
    )

  DocumentedNavBar.propTypes = {
    id: PropTypes.string
      .description('The DOM id attribute value to use for the element.')
      .defaultValue(false),
  }

  return DocumentedNavBar
}
