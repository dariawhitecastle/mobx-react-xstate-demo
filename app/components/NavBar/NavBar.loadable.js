/**
 * Asynchronously loads the component for NavBar
 */

import React from 'react'
import loadable from 'utils/loadable'
import LoadingIndicator from 'components/LoadingIndicator'

export default loadable(() => import('./NavBar'), {
  fallback: <LoadingIndicator />,
})
