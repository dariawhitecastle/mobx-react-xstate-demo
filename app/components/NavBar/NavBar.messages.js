/*
 * NavBar Messages
 *
 * This contains all the text for the NavBar component.
 */

import { defineMessages } from 'react-intl'

export const scope = 'app.components.NavBar'

export default defineMessages({
  authenticatedPrimary: {
    id: `${scope}.authenticatedPrimary`,
    defaultMessage: 'Authenticated Primary',
    buttons: [
      { title: 'Dashboard', to: '/dashboard' },
      { title: 'Assess', to: '/assess' },
      { title: `To-Do's`, to: '/todos' },
      { title: 'Critical Data', to: '/critical-data' },
      { title: `Guides & Resources`, to: '/guides-and-resources' },
      { title: 'Community', to: '/community' },
    ],
  },
  authenticatedPrimaryNonPremium: {
    id: `${scope}.authenticatedPrimaryNonPremium`,
    defaultMessage: 'Authenticated NonPremium Primary',
    buttons: [
      { title: 'Dashboard', to: '/dashboard' },
      { title: 'Assess', to: '/assess' },
      { title: `Guides & Resources`, to: '/guides-and-resources' },
      { title: 'Community', to: '/community' },
      { title: 'Upgrade', to: '/upgrade' },
    ],
  },
  authenticatedSecondary: {
    id: `${scope}.authenticatedSecondary`,
    defaultMessage: 'Authenticated Secondary',
    buttons: [
      { title: 'Profile', to: '/my-profile' },
      { title: 'Account', to: '/my-account' },
      { title: 'Help', to: '/help' },
    ],
  },
  unauthenticatedSecondary: {
    id: `${scope}.unauthenticatedSecondary`,
    defaultMessage: 'Unauthenticated Secondary',
    buttons: [
      { title: 'Help', to: '/help' },
      { title: 'Login', to: '/login' },
      { title: 'Create Account', to: '/create-account' },
    ],
  },
})
