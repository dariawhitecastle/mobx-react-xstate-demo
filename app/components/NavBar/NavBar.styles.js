import styled from 'styled-components'
import { NavLink } from 'react-router-dom'

import colors from '../../utils/colors'
import { Text } from '../Text'

export const StyledText = styled(Text).attrs(props => ({ ...props }))`
  font-size: ${props => props.fontSize || '16px'};
  font-family: 'Bryant Pro';
`

export const StyledLink = styled(NavLink).attrs(props => ({ ...props }))`
  color: ${colors.primaryText};
  text-decoration: none;

  &:focus,
  &:hover {
    text-decoration: underline;
  }

  &:visited,
  &:link,
  &:active {
    color: ${colors.primaryText};
  }
`
