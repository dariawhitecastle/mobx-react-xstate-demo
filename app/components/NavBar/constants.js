const NAV_HEIGHT = '80px'

const FORM_GRID_AREAS = {
  xsmall: [
    { name: 'logo', start: [0, 0], end: [3, 0] },
    { name: 'toggle', start: [8, 0], end: [9, 0] },
  ],
  small: [
    { name: 'logo', start: [0, 0], end: [3, 0] },
    { name: 'toggle', start: [9, 0], end: [9, 0] },
  ],
  medium: [
    { name: 'logo', start: [0, 0], end: [3, 0] },
    { name: 'toggle', start: [9, 0], end: [9, 0] },
  ],
  large: [
    { name: 'logo', start: [0, 0], end: [1, 0] },
    { name: 'center', start: [1, 0], end: [5, 0] },
    { name: 'right', start: [7, 0], end: [9, 0] },
  ],
}
const FORM_GRID_COLUMNS = ['10%', '10%', '10%', '10%', '10%', '10%', '10%', '10%', '10%', '10%']
const FORM_GRID_ROWS = [NAV_HEIGHT, 'max-content', 'max-content', 'max-content']

export { NAV_HEIGHT, FORM_GRID_AREAS, FORM_GRID_COLUMNS, FORM_GRID_ROWS }
