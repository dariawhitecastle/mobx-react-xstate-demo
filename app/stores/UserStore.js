import { action, observable } from 'mobx'
import { ignore } from 'mobx-sync'
import { createContext } from 'react'

export class UserStore {
  @observable user = {}

  @observable tempUser = {}

  @observable token = ''

  @ignore
  @observable
  isLoading = false

  @ignore
  @observable
  message = ''

  @ignore
  @observable
  error = ''

  @observable message = ''

  @action.bound
  setCurrentUser({ token = '', ...user } = {}) {
    this.token = token
    this.user = {
      ...user,
    }
  }
}

export const user = new UserStore()
export const UserStoreContext = createContext(user)
