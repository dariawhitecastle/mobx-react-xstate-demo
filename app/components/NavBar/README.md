<!-- ! This is a generated file. To make changes, edit <Component>.doc.js ! -->
## undefined
A NavBar

## Usage

```javascript
import { NavBar } from 'components/NavBar';
<NavBar />
```

## Properties

**id**

The DOM id attribute value to use for the element.

```
string
```
  