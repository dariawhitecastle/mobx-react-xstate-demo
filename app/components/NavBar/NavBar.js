/* eslint-disable react/prop-types */
import React, { useContext } from 'react'
import { withRouter } from 'react-router-dom'
import { ResponsiveContext } from 'grommet/es6/contexts'
import { observer } from 'mobx-react'
import { concat, find } from 'lodash'

// Components
import { Box } from '../Box'
import { Grid } from '../Grid'
import { Header } from '../Header'
import { Menu } from '../Menu'

// Stores
import { UserStoreContext } from '../../stores/UserStore'

// Utils and messages
import helpers from '../../utils/helpers'
import { doc } from './NavBar.doc'
import messages from './NavBar.messages'
import { NAV_HEIGHT, FORM_GRID_AREAS, FORM_GRID_COLUMNS, FORM_GRID_ROWS } from './constants'

// Styles
import { StyledLink, StyledText } from './NavBar.styles'

/**
 * Based on our grid area settings from `constants.js`, check if a specific
 * section is currently being displayed
 *
 * @param {str} size Current screen size
 * @param {str} name Grid section to check display status of
 */
const isShown = (size, name) => find(FORM_GRID_AREAS[size], { name })

/**
 * NavBar
 */
const NavBar = observer(
  withRouter(({ history }) => {
    /** @todo remove premium hardcoded value here once available through API */
    const userStore = useContext(UserStoreContext)
    const size = useContext(ResponsiveContext)
    const renderButtons = (key, fontSize = 'medium') => {
      const buttons = messages[key].buttons.map(({ title, to }) => (
        <StyledText
          alignSelf="center"
          fontSize={fontSize}
          margin={{ horizontal: '20px' }}
          key={`${title}`}
        >
          <StyledLink activeStyle={{ textDecoration: 'underline', fontWeight: 'bold' }} to={to}>
            {title}
          </StyledLink>
        </StyledText>
      ))

      /** @todo Find a more elegant solution for this Logout button */
      if (key === 'authenticatedSecondary')
        buttons.push(
          <StyledText
            alignSelf="center"
            fontSize={fontSize}
            margin={{ horizontal: '20px' }}
            key="Logout"
            onClick={() => userStore.setCurrentUser({})}
          >
            <StyledLink to="/">Logout</StyledLink>
          </StyledText>,
        )

      return buttons
    }

    const renderMenuLinks = key => {
      const links = messages[key].buttons.map(({ title, to }) => ({
        label: title,
        onClick: () => history.push(to),
      }))

      /** @todo Find a more elegant solution for this Logout button */
      if (key === 'authenticatedSecondary') {
        links.push({
          label: 'Logout',
          onClick: () => userStore.setCurrentUser({}),
        })
      }

      return links
    }

    /**
     * Check if the user is authenticated, if so, show the authenticated NavBar items.
     * Otherwise show the default items.
     */
    return (
      <Grid
        as={Box}
        areas={FORM_GRID_AREAS}
        columns={FORM_GRID_COLUMNS}
        data-testid="nav-bar"
        elevation="xsmall"
        rows={FORM_GRID_ROWS}
        style={{ height: NAV_HEIGHT }}
      >
        {/* Logo */}
        {isShown(size, 'logo') && (
          <Box gridArea="logo" justify="center" margin={{ horizontal: 'large' }}>
            <Header level={3}>PAR</Header>
          </Box>
        )}

        {/* Authenticated */}
        {userStore.token && (
          <>
            {isShown(size, 'center') && (
              <Box gridArea="center" direction="row">
                {renderButtons('authenticatedPrimary')}
              </Box>
            )}

            {isShown(size, 'right') && (
              <Box
                gridArea="right"
                direction="row"
                style={{ justifySelf: 'end' }}
                margin={{ right: '20px' }}
              >
                {renderButtons('authenticatedSecondary', 'small')}
              </Box>
            )}

            {isShown(size, 'toggle') && (
              <Box gridArea="toggle" justify="center" margin={{ horizontal: 'medium' }}>
                <Menu
                  a11yTitle="Menu"
                  dropAlign={{ top: 'bottom', right: 'right' }}
                  label=""
                  items={concat(
                    renderMenuLinks('authenticatedPrimary'),
                    renderMenuLinks('authenticatedSecondary'),
                  )}
                />
              </Box>
            )}
          </>
        )}

        {/* Unauthenticated */}
        {!userStore.token && (
          <>
            {isShown(size, 'right') && (
              <Box
                direction="row"
                gridArea="right"
                style={{ justifySelf: 'end' }}
                margin={{ right: '20px' }}
              >
                {renderButtons('unauthenticatedSecondary')}
              </Box>
            )}

            {isShown(size, 'toggle') && (
              <Box gridArea="toggle" justify="center" margin={{ horizontal: 'medium' }}>
                <Menu
                  a11yTitle="Menu"
                  dropAlign={{ top: 'bottom', right: 'right' }}
                  label=""
                  items={renderMenuLinks('unauthenticatedSecondary')}
                />
              </Box>
            )}
          </>
        )}
      </Grid>
    )
  }),
)

export default helpers.createWithDoc({
  docFunction: doc,
  component: NavBar,
})
