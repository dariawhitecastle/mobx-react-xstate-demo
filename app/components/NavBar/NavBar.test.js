/**
 *
 * Tests for NavBar
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react'
import { render, waitForElement } from '@testing-library/react'
import { IntlProvider } from 'react-intl'
import { MemoryRouter } from 'react-router-dom'
import { Grommet } from 'grommet'
import { ResponsiveContext } from 'grommet/es6/contexts'
import 'jest-dom/extend-expect'

import theme from '../../utils/theme'
import { UserStore, UserStoreContext } from '../../stores/UserStore'
import { DEFAULT_LOCALE } from '../../i18n'
import NavBar from './NavBar'

describe('<NavBar/>', () => {
  let userStore = null

  /** @todo test additional screen sizes */
  const renderComponent = renderUserStore =>
    render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <Grommet theme={theme}>
          <MemoryRouter initialEntries={['/']}>
            <UserStoreContext.Provider value={renderUserStore}>
              <ResponsiveContext.Provider value="large">
                <NavBar />
              </ResponsiveContext.Provider>
            </UserStoreContext.Provider>
          </MemoryRouter>
        </Grommet>
      </IntlProvider>,
    )

  beforeAll(() => {
    userStore = new UserStore()
  })

  it('renders and matches snapshot', () => {
    const { container } = renderComponent(userStore)

    expect(container).toMatchSnapshot()
  })

  it('does not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'log')

    renderComponent(userStore)

    expect(spy).not.toHaveBeenCalled()
  })

  it('displays the Create Account and Login links when unauthenticated', async () => {
    const { getByTestId } = renderComponent(userStore)
    const navBar = await waitForElement(() => getByTestId('nav-bar'))

    expect(navBar).toHaveTextContent('Create Account')
    expect(navBar).toHaveTextContent('Login')
  })

  it('displays the Logout link when authenticated', async () => {
    userStore.token = 'faketoken'
    const { getByTestId } = renderComponent(userStore)
    const navBar = await waitForElement(() => getByTestId('nav-bar'))

    expect(navBar).toHaveTextContent('Logout')
  })
})
