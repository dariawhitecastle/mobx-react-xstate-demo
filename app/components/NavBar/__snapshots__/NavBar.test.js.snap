// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavBar/> renders and matches snapshot 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 jjKgqr"
  >
    <div
      class="StyledGrid-sc-1wofa1l-0 dfkkNd StyledBox-sc-13pk1d4-0 gYVBxi"
      data-testid="nav-bar"
      style="height: 80px;"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 bbkGYr"
      >
        <h3
          class="StyledHeading-sc-1rdh4aw-0 ftOecI"
        >
          PAR
        </h3>
      </div>
      <div
        class="StyledBox-sc-13pk1d4-0 gdzbcl"
        style="justify-self: end;"
      >
        <span
          class="NavBarstyles__StyledText-cuj0qf-0 bTNZEJ StyledText-sc-1sadyjn-0 cAYdSQ"
          font-size="medium"
        >
          <a
            class="NavBarstyles__StyledLink-cuj0qf-1 gnewVS"
            href="/help"
            theme="[object Object]"
          >
            Help
          </a>
        </span>
        <span
          class="NavBarstyles__StyledText-cuj0qf-0 bTNZEJ StyledText-sc-1sadyjn-0 cAYdSQ"
          font-size="medium"
        >
          <a
            class="NavBarstyles__StyledLink-cuj0qf-1 gnewVS"
            href="/login"
            theme="[object Object]"
          >
            Login
          </a>
        </span>
        <span
          class="NavBarstyles__StyledText-cuj0qf-0 bTNZEJ StyledText-sc-1sadyjn-0 cAYdSQ"
          font-size="medium"
        >
          <a
            class="NavBarstyles__StyledLink-cuj0qf-1 gnewVS"
            href="/create-account"
            theme="[object Object]"
          >
            Create Account
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;
